src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsresettlementpassportapi/service/AppointmentsApiService.kt

```diff
@@ -0,0 +1,125 @@
+package uk.gov.justice.digital.hmpps.hmppsresettlementpassportapi.service
+
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.flow.toList
+import org.hibernate.query.sqm.tree.SqmNode.log
+import org.springframework.http.HttpStatus
+import org.springframework.stereotype.Service
+import org.springframework.web.reactive.function.client.WebClient
+import org.springframework.web.reactive.function.client.awaitBodyOrNull
+import uk.gov.justice.digital.hmpps.hmppsresettlementpassportapi.config.NoDataWithCodeFoundException
+import uk.gov.justice.digital.hmpps.hmppsresettlementpassportapi.config.ResourceNotFoundException
+import uk.gov.justice.digital.hmpps.hmppsresettlementpassportapi.data.Address
+import uk.gov.justice.digital.hmpps.hmppsresettlementpassportapi.data.Appointment
+import uk.gov.justice.digital.hmpps.hmppsresettlementpassportapi.data.AppointmentsList
+import uk.gov.justice.digital.hmpps.hmppsresettlementpassportapi.data.deliusapi.AppointmentDelius
+import uk.gov.justice.digital.hmpps.hmppsresettlementpassportapi.data.deliusapi.AppointmentsDeliusList
+import uk.gov.justice.digital.hmpps.hmppsresettlementpassportapi.jpa.repository.PrisonerRepository
+import java.time.LocalDate
+import java.time.OffsetDateTime
+import java.time.format.DateTimeFormatter
+
+@Service
+class AppointmentsApiService(
+  private val rpDeliusWebClientCredentials: WebClient,
+  private val prisonerRepository: PrisonerRepository,
+  private val communityApiService: CommunityApiService,
+) {
+
+  private fun fetchAppointments(crn: String, startDate: LocalDate, endDate: LocalDate, page: Int, size: Int): Flow<AppointmentsDeliusList> = flow {
+    val data = rpDeliusWebClientCredentials.get()
+      .uri(
+        "/appointments/{crn}?page={page}&size={size}&startDate={startDate}&endDate={endDate}",
+        mapOf(
+          "crn" to crn,
+          "size" to size,
+          "page" to page,
+          "startDate" to startDate.toString(),
+          "endDate" to endDate.toString(),
+        ),
+      )
+      .retrieve().onStatus({ it == HttpStatus.NOT_FOUND }, { throw ResourceNotFoundException("CRN  $crn not found") })
+    val pageOfData = data.awaitBodyOrNull<AppointmentsDeliusList>()
+    if (pageOfData != null) {
+      emit(pageOfData)
+    }
+  }
+
+  suspend fun getAppointmentsByNomisId(
+    nomisId: String,
+    startDate: LocalDate,
+    endDate: LocalDate,
+    pageNumber: Int,
+    pageSize: Int,
+  ): AppointmentsList {
+    if (nomisId.isBlank() || nomisId.isEmpty()) {
+      throw NoDataWithCodeFoundException("Prisoner", nomisId)
+    }
+
+    if (pageNumber < 0 || pageSize <= 0) {
+      throw NoDataWithCodeFoundException(
+        "Data",
+        "Page $pageNumber and Size $pageSize",
+      )
+    }
+    val prisonerEntity = prisonerRepository.findByNomsId(nomisId)
```

Comment:
I think this may be inconsistent with the approach elsewhere. Usually if we need a prisoner to be in the database and they are missing then we throw a 404 not found. Maybe we need to rethink this as it mean you have to call the GET on the prisoner first but in our app that works fine for now.